fronted.go

```diff
@@ -139,42 +207,72 @@ func Vet(m *Masquerade, pool *x509.CertPool, testURL string) bool {
 	return masq.postCheck(conn, testURL)
 }
 
-// findWorkingMasquerades finds working masquerades by vetting them in batches and in
-// parallel. Speed is of the essence here, as without working masquerades, users will
+// findWorkingFronts finds working domain fronts by vetting them in batches and in
+// parallel. Speed is of the essence here, as without working fronts, users will
 // be unable to fetch proxy configurations, particularly in the case of a first time
 // user who does not have proxies cached on disk.
-func (f *fronted) findWorkingMasquerades(listener func(f *fronted)) {
-	// vet masquerades in batches
+func (f *fronted) findWorkingFronts() {
+	// vet fronts in batches
 	const batchSize int = 40
-	var successful atomic.Uint32
 
-	// We loop through all of them until we have 4 successful ones.
-	for i := 0; i < len(f.masquerades) && successful.Load() < 4; i += batchSize {
-		f.vetBatch(i, batchSize, &successful, listener)
+	// Keep looping through all fronts making sure we have working ones.
+	i := 0
+	for {
+		// Continually loop through the fronts in batches until we have 4 working ones,
+		// always looping around to the beginning if we reach the end.
+		// This is important, for example, when the user goes offline and all fronts start failing.
+		// We want to just keep trying in that case so that we find working fronts as soon as they
+		// come back online.
+		if f.connectingFronts.size() < 4 {
+			f.vetBatch(i, batchSize)
+			i = index(i, batchSize, f.frontSize())
+		} else {
+			select {
+			case <-f.stopCh:
+				log.Debug("Stopping parallel dialing")
+				return
+			case <-time.After(time.Duration(rand.IntN(12000)) * time.Millisecond):
+			}
+		}
```

Comment:
This is also a key change @garmr-ulfr

Reply:
Ah interesting so make sure x goroutines are always running...gotta think about that but I like the idea

Reply:
The idea of just using workers instead of a waitgroup is interesting. My only hesitation is that the pattern tends to be that, if a front is going to fail, it's usually going to timeout, in which case the entire batch just fails at the 5 second mark. That said, that's not always the case, and sometime they fail right away on a cert mismatch or soemthing, so I do agree with the change -- will look more at that today